### DISCUSSION TITLE
Request changes in the pwm frequency on a esp32

### DISCUSSION THREAD

**Discussion by Baluar81** [arendst/Tasmota#13123] [Ideas]
Hi there!

Firts, thank you for your project. It's wonderful.

I would like to use ssr relay with esp32 controlling AC's complete semicycles with the 50 Hz signal.

For example: 

50% 1 semicycle in on mode and 1 off  PWM Frecuency  50Hz
10% 1 semicycle in on and 9 in off PWM frequency 10Hz
5% 1 semicycle in on and 10 in off PWM frequency 5Hz

I have read tha in a tasmoted  esp32 the PWM frequency is between 40kHz and 50 KHz, https://tasmota.github.io/docs/Commands/

> 
PwmFrequency | 1 = reset PWM frequency to 223Hz40..4000 or 40..50000 = set PWM frequency (40Hz to 4kHz on ESP 82xx / 40-50kHz on ESP32)As of v8.3.0 the default frequency changed to 977Hz

I ask you if possible to change the PWM frequency on a esp32 to 4Hz-50Khz in future releases.

Thanks in advance.

**Comment by sfromis** [comment]
That minimum of 40 Hz is not up to Tasmota, changing it would have to be supported in Arduino Core.

**Comment by s-hadinger** [comment]
As per this https://github.com/espressif/esp-idf/issues/2255 the minimum frequency is 15Hz in hardware. Going below needs to rewrite low level functions, which probably won't happen.

**Reply by sfromis** [reply]
I suppose that it could be feasible to control 4 Hz PWM signals from Berry, if rock stable timing is not required.

**Comment by s-hadinger** [comment]
Moreover your need for a succession of patterns won't work with pwm. Maybe RMT is better suited although I'm not sure it can go that low in frequency

**Comment by s-hadinger** [comment]
Well not quite. The pulse is still 10mS followed by 190mS off. That's possible but not ideal

**Comment by barbudor** [comment]
As he is referring to % of cycle, I would assume this is synced to zero-crossing
This has been discussed multiple time for speed control of AC motors and rejected because it is proven to be a wrong way to control AC motor's speed.

@Baluar81 Please advise your use case.

**Reply by Baluar81** [reply]
Thank for your responses.

I'm using (and working) an esp32 with micropython to control by a ssr with %'s (with the frequencies i mentioned in my firts post) of AC cycle in a photovoltaic installation and use his excedents with an electrical water heater (or any kind of electrical resistance charge, like an oven) . When i saw it's possible to put Tasmota on a esp32 i suppuse if possible to change the frequency like I use with mycropython.

Thaks again!

**Comment by barbudor** [comment]
@Baluar81 So as per the discussion above, it could be possible to reduce the minimum frequency to 15Hz on ESP32.

However what you are describing:
> 50% 1 semicycle in on mode and 1 off PWM Frecuency 50Hz
10% 1 semicycle in on and 9 in off PWM frequency 10Hz
5% 1 semicycle in on and 10 in off PWM frequency 5Hz

This is not PWM but variable frequency and it is not supported neither by ESP PWM hardware nor Tasmota.
To work properly on semi cycle, this requires to be synced to the AC cycle with a zero-crossing detection. 
Are you using zero-crossing detection in your micropython solution ?

**Comment by Baluar81** [comment]
Hi again.

The ssr i use to feed the electrical water heater includes a zero-crossing detection and is like that:
https://es.aliexpress.com/item/32841699288.html

From esp 32 gnd to - and 12,13 or 14 to +.

It works perfectly with micropython in the frequencies I showed you and changes PWM duty;

50%... frequency  50hz and duty 50%

10% ... 10 hz and duty 10%

80%... 20 hz and duty 80%

I assure it's working perfectly in micropython :)

**Reply by s-hadinger** [reply]
Thanks for the clarification. Now I get it. What method are you using with micropython? Standard pwm lib?

**Reply by s-hadinger** [reply]
I will do the math tomorrow, but it looks like PWM can go way lower than 15Hz...

**Comment by Baluar81** [comment]
Hi there again.

I have been talking with the person who coded de micropython script and he told me so. It used stansard PWM lib.

**Reply by s-hadinger** [reply]
We owe you an apology. You can indeed go as low as 2Hz. I will change the limit in the code. However changing the frequency on the fly is not supported right now and behaves erratically. I suppose that you want it to change on the fly too?

**Reply by s-hadinger** [reply]
Ok, it does work but you need to set off first, change the frequency and turn it on again in a row. Changing the frequency will reset all PWMs to off anyways, but you need to change the value otherwise Tasmota commands will think that the value didn't change

**Reply by barbudor** [reply]
@Baluar81 can you please provide a link to this script?
I'm convinced the problem is not a PWM problem

**Comment by Baluar81** [comment]
There is nothing to apologice, @s-hadinger, :) We are learning some new stuff every day :D. I'have asked the coder about changin g PWM frenquency in on the fly, but i'm still wating an answer, However I suppouse there could be necessary.

@barbudor, in this link in sourceforge https://sourceforge.net/projects/pvcontrol/ there is the programs we use to monitoring and controlling PV installations, included the PWM with an esp32/NodeMCU and a ssr's with semicycles and phase angle.

When i have an answer i wil post again.

**Comment by Baluar81** [comment]
I'm here again!

At least, the person who code the script has told me that every time we send a change in the potence, changes PWM's frequency and duty too...

I hope this can help us... :)

**Comment by s-hadinger** [comment]
It should work fine with the latest commit

**Reply by Baluar81** [reply]
This is the code:

https://docs.micropython.org/en/latest/esp8266/tutorial/pwm.html

It's for an esp8266 but works perfectly with an esp32. Every time esp32 receives an mqtt message with the necessary power, changes the PWM's frequency and duty  to adjust the PWM signal that it sends to the SSR relay control.


> It should work fine with the latest commit

Sorry, but my english is not well enough. Did you say that would work in next releases?

**Reply by Jason2866** [reply]
It does mean. The change is now in the actual development firmware binaries. You can get it via OTA http://ota.tasmota.com/tasmota32/ and http://ota.tasmota.com/tasmota/ or from https://github.com/arendst/Tasmota-firmware

**Reply by bob-u** [reply]
Would it be possible to enable lower frequencies on ESP8266?  I downloaded latest development for ESP8266 and it still insists on 40Hz minimum.  I am also trying to control SSR with it and need lower PWM frequencies.

**Reply by Jason2866** [reply]
No, since this would need changes in Arduino core Esp8266. You can use a ESP32.

**Reply by Weltkugel** [reply]
Hi,

thanks for development. I have got a PWM with 2hz working fine on a [Board](https://user-images.githubusercontent.com/45427770/154621619-bcb34f78-0423-4665-81c7-70e804298c67.jpg) with ESP32-WROOM-32E-module. On the [Wemos-S2mini-Board](https://www.wemos.cc/en/latest/s2/s2_mini.html) is a ESP32-S2FN4R2-module soldered and the PWM with 2hz **_is not working_**. The PWM-frequenz is clearly higher. Is a 2hz PWM also possible with the ESP32-S2 or are there other hardware limitations?

**Reply by stefanbode** [reply]
Not yet seen. I only know that esp32 frequency only allow integer. So 1 and 2 should work but it is not very flexible anyhow. This was much easier with esp8266. The only way on esp32 is working with interrupts. What did you require in detail?

**Reply by sfromis** [reply]
From a simple test (connecting a PWM pin to a Counter pin, I could not get ESP32-S2 PWM frequency below 39 Hz. With "original" ESP32, I got it down to 2 Hz.

if you do not need to manage duty cycle, but only frequency, you can have a "slow PWM" using the blink feature, timing in 1/10 second intervals. You could also manage such pulsing using a small Berry program.

**Reply by Weltkugel** [reply]
@ stefanbode 
Thanks for your answer. I required the confirmation, that there is a difference between "original" and "S2" variants of the ESP32, so I do not have to search any further ;-) Should we update the **PWMFrequency**-Wiki with this S2-information?

@ sfromis 
Thank you for the confirmation. I want to use photovoltaic excess for waterheating and have to manage the duty cycle. Here is a german description [www.photovoltaikforum.com](https://www.photovoltaikforum.com/thread/188108-balkon-pv-100-energie-nutzung-ww-erw%C3%A4rmung/). As it works with the original ESP32 im using this type.

**Comment by Baluar81** [comment]
Oh, thanks a lot.

In next days i will try and tell you how it works!

**Reply by stefanbode** [reply]
I just updated the zero-cross driver. See here one of the todays topics. This will operate dimmers with phase cut and zero detect. This is at the end more than complicated and tricky. But it works. I uses internal esp8266 manipulation of the Pwm frequency.

**Comment by cacengineering** [comment]
Thank you @stefanbode - have been following your work on pwm.  Not sure if part of this post or to start new thread - but how do you synchronise the start of the pwm cycle with the zero crossing ?  I am using counter for zero crossing detect of 100hz (rectified 50Hz).  Change in frequency would be great too but right now need to get the synchronisation correct to switch a thyristor.

**Reply by stefanbode** [reply]
One past I changed the frequency slightly based on measurement to keep it in sync. The new implementation also running on esp32 do it hardcore like ESPeasy and use high frequency interrupt to see if zero cross was executed. These interrupt can work on a quite high frequency. After you detect the zero cross you check on the time microseconds if it is time to switch On the relay. I moved away completely from using the pwm. There is now a dedicated ino file for the zero-crossing. This should explain a lot.

**Reply by cacengineering** [reply]
Thanks.  I started out planning on using Berry to measure when the zero cross occurred, and then trigger a timer to trigger the gate - but then saw the PWM function that looked really useful. Am trying to work out how to synch either PWM or timer to the counter pulses.  Are you running the high frequency interupt on ESP easy or Tasmota. If on Tasmota - how can I access it? What pin setting?

**Reply by stefanbode** [reply]
It is for sure Tasmota. Please search for zero-crossing in the documentation. Should answer all your questions. I assume you cannot do it in Berry.
